kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/ApiVisitor.java

```diff
@@ -0,0 +1,79 @@
+/**
+ * Copyright (C) 2015 Red Hat, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.fabric8.kubernetes.client;
+
+import io.fabric8.kubernetes.api.model.APIResource;
+import io.fabric8.kubernetes.api.model.GenericKubernetesResource;
+import io.fabric8.kubernetes.api.model.GenericKubernetesResourceList;
+import io.fabric8.kubernetes.client.dsl.MixedOperation;
+import io.fabric8.kubernetes.client.dsl.Resource;
+import io.fabric8.kubernetes.client.utils.ApiVersionUtil;
+
+/**
+ * Used to perform operations across all resources known to the api server.
+ */
+public interface ApiVisitor {
+
+  public enum ApiVisitResult {
+    /**
+     * If returned, stop the visiting process immediately
+     */
+    TERMINATE,
+    /**
+     * Continue visiting
+     */
+    CONTINUE,
+    /**
+     * Skip any visiting that would occur under this point
+     */
+    SKIP
+  }
+
+  /**
+   * Visit the api. The core/legacy group will be an empty string.
+   * 
+   * @param api the group name
+   * @return the result to control future actions. {@link ApiVisitResult#SKIP} will skip visiting all versions of this api.
+   */
+  default ApiVisitResult visitApi(String api) {
```

Comment:
Why was the word Api chosen?

Won't this be more confusing than calling it visitGroup?

Kubernetes coordinates:
- Group: API group the resource belongs to, or empty for the core resources
- Version
- Kind
- `apiVersion`: `$group/$version`
- Plural: HTTP resource name where GVK resources are exposed

Reply:
> Why was the word Api chosen?

All of the other client entry points related to "apis" are api methods - getApiGroups, getApiGroup, getApiResources, hasApiGroup, so it seems like we should keep consistency with the api term.  If you prefer to use group in this method name that's fine - visitApiGroup or just visitGroup.

Reply:
I understand that `apiGroup` -> `visitApiGroup` is what might both clearer and consistent, right?